### Python Fundamentals for Data Analysts

# 4.6: Combining & Exporting Data

# Part 1

### Content:

1. Data cleaning on "Order Prior"
2. Perform Merging between "Order" and "Order Prior"

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import os
from datetime import datetime

In [2]:
# Set path to access data files & Load data file
path = r'C:\Users\tsoew\OneDrive\Desktop\InstaCart Basket Analysis'

# Loading orders.csv
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_clean.csv'), index_col = False)
# Loading orders_products_prior.csv
df_ords_prior = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)


In [3]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [4]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

### Perform Data Cleaning on "Order Prior" data

In [6]:
df_ords_prior.describe()

order_id    product_id  add_to_cart_order     reordered
count  3.243449e+07  3.243449e+07       3.243449e+07  3.243449e+07
mean   1.710749e+06  2.557634e+04       8.351076e+00  5.896975e-01
std    9.873007e+05  1.409669e+04       7.126671e+00  4.918886e-01
min    2.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00
25%    8.559430e+05  1.353000e+04       3.000000e+00  0.000000e+00
50%    1.711048e+06  2.525600e+04       6.000000e+00  1.000000e+00
75%    2.565514e+06  3.793500e+04       1.100000e+01  1.000000e+00
max    3.421083e+06  4.968800e+04       1.450000e+02  1.000000e+00

In [7]:
df_ords_prior.isnull().sum()

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64

**Note:** No missing values

In [8]:
df_ords_prior[df_ords_prior.duplicated()]

Empty DataFrame
Columns: [order_id, product_id, add_to_cart_order, reordered]
Index: []

**Note:** No duplicates

In [9]:
df_ords_prior.shape

(32434489, 4)

**Note:** 
- Total rows: 32,434,489
- Total columns: 4

### Perform Merging between "Orders" and "Orders Prior"

In [10]:
# This will create INNER JOIN in a new dataframe, that contains the combined df_ords 
# and df_ords_prior dataframes and uses the “order_id” column as its key. It also includes 
# the indicator = True argument so that you can check for a full match.

df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [11]:
df_merged_large.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both  
5                    15.0         196                  1          1   both  
6                    15.0       10258                  2          0   both  
7                    15.0       12427                  3          1   both  
8                    15.0       13176                  4          0   both  
9                    15.0       26088                  5          1   both

In [12]:
df_merged_large.shape

(32434489, 10)

**Note:** Now we have additional 6 rows from the merging
- Total rows: 32,434,489
- Total columns: 10

In [13]:
# Check the _merge columns to indicate successful INNER MERGE
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [17]:
# Export data to pkl and put timing to know the difference between two methods

print ('Start time: ' + datetime.now().strftime("%H:%M:%S"))
df_merged_large.to_pickle(os.path.join(path, 'Data','Prepared Data', 'ord_prod_combined_1.pkl'))
print ('End time: ' + datetime.now().strftime("%H:%M:%S"))

Start time: 11:55:20
End time: 11:55:32


# End